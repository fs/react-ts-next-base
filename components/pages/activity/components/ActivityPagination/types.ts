import { Dispatch, SetStateAction } from 'react';

import { PageInfoFragment } from 'graphql/fragments/__generated__/pageInfo.generated';

export type TActivityPagination = {
  pageInfo: PageInfoFragment;
  setBeforeCursor: Dispatch<SetStateAction<string | undefined>>;
  setAfterCursor: Dispatch<SetStateAction<string | undefined>>;
  setPageNumber: Dispatch<SetStateAction<number>>;
  pageNumber: number;
};
