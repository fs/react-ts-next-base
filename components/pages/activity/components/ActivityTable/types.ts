import { ActivityFragment } from 'graphql/fragments/__generated__/activityInfo.generated';

export type TActivityTable = {
  activities: ActivityFragment[];
};
