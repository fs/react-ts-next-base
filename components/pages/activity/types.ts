import { MultiValue, SingleValue } from 'react-select';
import { ActivityEvent } from 'graphql/types';

export type TActivityOption = {
  value: ActivityEvent;
  label: string;
};

export type TAvtivitySizeOption = {
  value: number;
  label: string;
};

export type TEventChange = (
  value: MultiValue<TActivityOption> | SingleValue<TActivityOption> | null,
) => void;
export type TSizeChange = (
  value: MultiValue<TAvtivitySizeOption> | SingleValue<TAvtivitySizeOption> | null,
) => void;
