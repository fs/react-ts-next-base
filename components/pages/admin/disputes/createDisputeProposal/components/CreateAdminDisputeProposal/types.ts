import { OrderInfoFragment } from 'graphql/fragments/__generated__/orderInfo.generated';
import { DisputeFragment } from 'graphql/fragments/__generated__/disputeInfo.generated';

export type TCreateAdminDisputeProposal = {
  order: OrderInfoFragment;
  dispute: DisputeFragment;
};

export type TCurrentProposal = {
  dispute: DisputeFragment;
};
