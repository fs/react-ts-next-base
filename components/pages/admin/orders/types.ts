import { tabsConfig } from './constants';

export type TConfig = keyof typeof tabsConfig;
