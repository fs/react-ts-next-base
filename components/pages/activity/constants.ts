import { ActivityEvent } from 'graphql/types';
import { TActivityOption, TAvtivitySizeOption } from './types';

export const activityPageSizes = [5, 10, 25, 50];

export const activityPageSizeOptions: TAvtivitySizeOption[] = activityPageSizes.map(item => ({
  value: item,
  label: String(item),
}));

export const activityEvents: TActivityOption[] = [
  { value: ActivityEvent.UserLoggedIn, label: 'User logged in' },
  { value: ActivityEvent.UserRegistered, label: 'User registered' },
  { value: ActivityEvent.UserResetPassword, label: 'User reset password' },
  {
    value: ActivityEvent.ResetPasswordRequested,
    label: 'Reset password requested',
  },
  { value: ActivityEvent.UserUpdated, label: 'User updated' },
];
