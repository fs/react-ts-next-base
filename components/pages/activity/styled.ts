import styled from 'styled-components';

export const Wrapper = styled.div`
  position: relative;

  display: flex;
  flex-direction: column;

  width: 100%;
  padding-top: 1rem;
`;
