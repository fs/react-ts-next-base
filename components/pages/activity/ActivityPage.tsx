import { useState } from 'react';

import withAuth from 'lib/auth/withAuth';
import withAuthSecurity from 'lib/auth/withAuthSecurity';
import withGetDataFromTree from 'lib/apollo/withGetDataFromTree';

import { useMeActivities } from 'lib/apollo/hooks/state/meActivity';

import { ActivityEvent } from 'graphql/types';
import { TNextPage } from 'lib/apollo/types';

import Loader from 'components/shared/atoms/Loader';
import Select from 'components/shared/atoms/Select';
import DefaultTemplate from 'components/shared/templates/DefaultTemplate';
import ErrorPage from 'pages/_error';

import ActivityPagination from './components/ActivityPagination';
import ActivityTable from './components/ActivityTable';

import { ACTIVITY_EVENTS, ACTIVITY_PAGE_SIZE_OPTIONS, ACTIVITY_PAGE_SIZES } from './constants';

import { Wrapper } from './styled';
import { TEventChange, TSizeChange } from './types';

export const ActivityPage: TNextPage = () => {
  const [beforeCursor, setBeforeCursor] = useState<undefined | string>();
  const [afterCursor, setAfterCursor] = useState<undefined | string>();
  const [activityEvent, setActivityEvent] = useState<undefined | ActivityEvent>();
  const [pageNumber, setPageNumber] = useState(1);
  const [pageSize, setPageSize] = useState(ACTIVITY_PAGE_SIZES[0]);

  const { activities, pageInfo, loading, error } = useMeActivities({
    before: beforeCursor,
    after: afterCursor,
    event: activityEvent,
    pageSize,
  });

  const resetState = () => {
    setBeforeCursor(undefined);
    setAfterCursor(undefined);
    setPageNumber(1);
  };

  const onEventChange: TEventChange = value => {
    if (value === null) {
      setActivityEvent(undefined);
      resetState();
    } else if (value && 'value' in value) {
      setActivityEvent(value.value);
      resetState();
    }
  };

  const onPageSizeChange: TSizeChange = value => {
    if (value === null) {
      setActivityEvent(undefined);
      resetState();
    } else if (value && 'value' in value) {
      setPageSize(value.value);
      resetState();
    }
  };

  if (!loading && (error || !activities)) return <ErrorPage statusCode={404} />;

  return (
    <DefaultTemplate>
      <Wrapper>
        <Select
          name="activity-event"
          title="Choose activity event:"
          placeholder="Choose activity event:"
          options={ACTIVITY_EVENTS}
          value={ACTIVITY_EVENTS.find(({ value }) => value === activityEvent) || null}
          onChange={onEventChange}
          $mb={16}
        />
        <Select
          name="activity-size"
          title="Choose activity page size:"
          options={ACTIVITY_PAGE_SIZE_OPTIONS}
          value={ACTIVITY_PAGE_SIZE_OPTIONS.find(({ value }) => value === pageSize) || null}
          onChange={onPageSizeChange}
          isClearable={false}
          $mb={16}
        />

        {pageInfo && (
          <ActivityPagination
            pageInfo={pageInfo}
            setBeforeCursor={setBeforeCursor}
            setAfterCursor={setAfterCursor}
            setPageNumber={setPageNumber}
            pageNumber={pageNumber}
          />
        )}

        {!loading && activities ? (
          <ActivityTable activities={activities} />
        ) : (
          <Loader testId="activity-loading" />
        )}
      </Wrapper>
    </DefaultTemplate>
  );
};

export default withGetDataFromTree(withAuth(withAuthSecurity(ActivityPage)));
