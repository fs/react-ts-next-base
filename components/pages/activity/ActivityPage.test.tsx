import React from 'react';
import { render, screen } from '@testing-library/react';

import renderWithTheme from '__tests__/helpers/renderWithTheme';
import renderWithApolloClient from '__tests__/helpers/renderWithApolloClient';

import { useActivities } from 'lib/apollo/hooks/state/activity';

import { TPageProps } from 'lib/apollo/types';
import { mockUseActivitiesData } from '__tests__/mocks/mockActivities';

import { ActivityPage } from './ActivityPage';

jest.mock('lib/apollo/hooks/state/activity');

const mockedUseActivities = useActivities as jest.MockedFunction<typeof useActivities>;

describe('Activity page', () => {
  test('should render correctly', () => {
    // Arrange
    const mockUseActivity = jest.fn(() => mockUseActivitiesData);
    mockedUseActivities.mockImplementation(mockUseActivity);

    // Act
    render(
      renderWithTheme(
        renderWithApolloClient((pageProps: TPageProps) => <ActivityPage {...pageProps} />),
      ),
    );
    const pageContent = screen.getByTestId('activity-table');

    // Assert
    expect(pageContent).toMatchSnapshot();
  });

  test('should show loader while loading', () => {
    // Arrange
    const mockUseActivity = jest.fn(() => ({ ...mockUseActivitiesData, loading: true }));
    mockedUseActivities.mockImplementation(mockUseActivity);

    // Act
    render(
      renderWithTheme(
        renderWithApolloClient((pageProps: TPageProps) => <ActivityPage {...pageProps} />),
      ),
    );
    const loadingContent = screen.getByTestId('activity-loading');

    // Assert
    expect(loadingContent).toMatchSnapshot();
  });
});
