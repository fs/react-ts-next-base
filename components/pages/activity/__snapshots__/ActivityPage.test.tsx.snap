// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity page should render correctly 1`] = `
<table
  class="sc-dIouRR jbjBFL"
  data-testid="activity-table"
>
  <thead>
    <tr>
      <th
        class="sc-dmRaPn oPUDT"
        colspan="2"
      >
        Title
      </th>
      <th
        class="sc-dmRaPn oPUDT"
        colspan="1"
      >
        Description
      </th>
      <th
        class="sc-dmRaPn oPUDT"
        colspan="1"
      >
        Date
      </th>
      <th
        class="sc-dmRaPn oPUDT"
        colspan="1"
      >
        User
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      data-testid="activity-row"
    >
      <td
        class="sc-hHLeRK AxKNg"
        color="#00e676"
      />
      <td
        class="sc-bBrHrO kdXtnc"
      >
        <div
          class="sc-ivTmOn kDpbig"
        >
          title
        </div>
      </td>
      <td
        class="sc-bBrHrO kdXtnc"
      >
        <div
          class="sc-ivTmOn kDpbig"
        >
          text
        </div>
      </td>
      <td
        class="sc-bBrHrO kdXtnc"
      >
        <div
          class="sc-ivTmOn kDpbig"
        >
          Invalid Date
        </div>
      </td>
      <td
        class="sc-bBrHrO kdXtnc"
      >
        <div
          class="sc-ivTmOn kDpbig"
        >
          <div
            class="sc-kDDrLX fNsyKy"
          >
            <img
              alt="Avatar"
              class="sc-iqcoie gWfHcr"
              src="/images/avatar-placeholder.png"
            />
          </div>
          <span
            class="sc-kgflAQ ccjRmf"
          >
            firstName lastName
             (
            test@test.com
            )
          </span>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Activity page should show loader while loading 1`] = `
<div
  class="sc-llJcti cgvikk"
  data-testid="activity-loading"
>
  <div
    class="sc-iIPllB bAGyCK"
    size="45"
  >
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;
